### PR TITLE
Add async mode and mesh tensor support to reduce scatter op

### PR DESCRIPTION
**Pull request by SeanNijjar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/10584)

### Problem description
These features were missing for the op

### What's changed
Added support for async mode and mesh tensor for reduce scatter

### Checklist
- [x] Post commit CI passes
  - https://github.com/tenstorrent/tt-metal/actions/runs/10133652053
- [x] T3000 Frequent passes
  - https://github.com/tenstorrent/tt-metal/actions/runs/10104717859
- [x] T3000 demos passes
  - https://github.com/tenstorrent/tt-metal/actions/runs/10104725899
- [x] T3000 unit tests passes
  - https://github.com/tenstorrent/tt-metal/actions/runs/10104733482
- [x] T3000 model perf
  - https://github.com/tenstorrent/tt-metal/actions/runs/10133649213
- [x] New/Existing tests provide coverage for changes
  - Enabled async mode in tests. Updated reduce scatter tests to use mesh tensor

### CODE REVIEW

**Comment by SeanNijjar** [comment]
FYI @djordje-tt

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_nightly.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_reduce_scatter_nightly.py:51]: should this remain true?

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_nightly.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_reduce_scatter_nightly.py:61]: Why do we turn it off by default?

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py:120]: mm, do we have a fully duplicated code in two places?

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:3]: remove

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:11]: can we move all such files to ccl/common in some future PR?

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: this file should be named 
ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter.hpp

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: and I'd prefer `reduce_scatter` method to be gone from 
ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.hpp
and its implementation to be in reduce_scatter.cpp

also not sure why we need reduce_scatter_impl

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:85]: You use next code to register this op
```
constexpr auto reduce_scatter =
    ttnn::register_operation<"ttnn::reduce_scatter", ttnn::operations::ccl::ExecuteReduceScatter>();
```

This will decorate an operation and will automatically wrap the call into a launch_op.
Can we remove launch_op from here?


see https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/decorators.hpp#L221

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:11]: yes. I'm planning on moving gtests on Friday so I'll move this at the same time. I also have some reorg to do with some of these headers as well (they need to be split up as right now they've become sort of accumulation points for separate classes components).

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py
Comment (SeanNijjar) [reply tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py:120]: will commonize

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: renamed file to reduce_scatter.hpp but haven't applied this other change. When I apply it as I understand it, there are some issues with device/tensor identification  that lead to a failure. I'll investigate it tomorrow.

Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: I dug into the change a little where we remove the explicit launch op and instead register the op with `register_operation_with_auto_launch_op`. Unfortunately this leads to further issues which haven't been root caused yet. 

Some of this seems to be due to missing codepaths for support for `MultiDeviceStorage` storage types in some of the ttnn operations infra. I added these paths where I suspected they were missing (obviously pending review) to unblock and make further progress. 

However, I hit another issue that seems to be taking me down a rabbit hole. At this point I am going to open an issue  to enable this functionality for CCL ops with mesh tensors. I don't think this has been done yet and so I believe I'm path clearing some unrelated (to reduce scatter specifically) issues in the general op infra.

The next issue I am seeing is that querying the mesh tensor for workers returns the correct number of workers for the (I assume) first chip but for the second (and beyond) chips where we try to generate a reduce scatter program, we are only getting a single worker, so we aren't able to identify our neighbour device IDs (in this case in the multi-device ring).

So I'm going to push back and say that I will for now stay with using the explicit `LaunchOp` approach and perhaps let's revisit this when some cycles open up.

Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: Issue tracked here: https://github.com/tenstorrent/tt-metal/issues/10756

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:8]: resolved as its a separate work item that I'll start shortly

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/reduce_scatter_op.hpp:14]: merged reduce_scatter_impl + reduce_scatter.

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp:82]: See earlier comment and new issue tracking the problems: https://github.com/tenstorrent/tt-metal/issues/10756

**Comment by SeanNijjar** [comment]
FYI @cfjchu - I had to update one other place that had an incorrect assert on the variant alternative it was expecting. I added a corresponding gtest for that change as well, like how you did with your prior change (it's the latest commit on the stack). All these'll get squashed before merge.

**Comment by SeanNijjar** [comment]
FYI @johanna-rock-tt and @djordje-tt, I know you are updating falcon 40b to support async and (I believe) mesh tensor. The new reduce scatter changes now only support the mesh tensor way of calling the op. 

Because of that, I've made localized changes to the falcon mlp file where I essentially wrap the reduce scatter call such that I pack the input tensor list into a mesh tensor as an argument, and then unpack the returned mesh tensor back to a tensor list. I'm running the model perf suite to make sure this doesn't detrimentally affect performance. 

I'd like your blessing on this temporary change so I can get the reduce scatter updates in and you can remove these conversions when you go to full mesh tensor migration.

**Review by djordje-tt — APPROVED** [review]

**Comment by cfjchu** [comment]
> FYI @cfjchu - I had to update one other place that had an incorrect assert on the variant alternative it was expecting. I added a corresponding gtest for that change as well, like how you did with your prior change (it's the latest commit on the stack). All these'll get squashed before merge.

sounds good, thanks!

**Review by cfjchu — APPROVED** [review]

**Pull request merged**
